Checker_TestFunctions.hpp

```diff
@@ -6,41 +6,87 @@
 #include "Standard_Battery_Specifications.hpp"
 using namespace std;
 
+const int Tolerance_percentage = 5;
 
-inline bool Combined_Check(bool chargetemp, bool SOC, bool Charge_Rate){
+inline bool ActionIsRequired(bool chargetemp, bool SOC, bool Charge_Rate){
     bool Check_m = (chargetemp && SOC && Charge_Rate);
     return Check_m;
 }
 
-template <typename temperature, class Charge_Discharge_SOC_Temp>
-bool ChargeTemp_SOC_CRate_Check(temperature temp_SOC, Charge_Discharge_SOC_Temp ClsName) {
+template <class Threshold_Tolerance>
+inline int Warning_Tolerance(Threshold_Tolerance Thres_tolerance_Cls) {
+    float temp = ((Thres_tolerance_Cls.Upper_Threshold) * (Tolerance_percentage)) / 100;
+    return (temp);
+}
+
+template <typename test_thres, class Charge_Discharge_SOC_Temp>
+void IsItApproaching_LowerLimit(test_thres test_treshold, Charge_Discharge_SOC_Temp ClsName, float Warning_tolerance_range) {
+
+    if(test_treshold <= (ClsName.Lower_Threshold + Warning_tolerance_range) && test_treshold >= ClsName.Lower_Threshold) {
+        cout << "Warning: Approaching Lower Limit\n";
+    }
+}
+
+template <typename test_thres, class Charge_Discharge_SOC_Temp>
+void IsItApproaching_UpperLimit(test_thres test_treshold, Charge_Discharge_SOC_Temp ClsName, float Warning_tolerance_range) {
+
+    if (test_treshold <= (ClsName.Upper_Threshold) && test_treshold >= (ClsName.Upper_Threshold  - Warning_tolerance_range))
+    {
+        cout << "Warning: Approaching Upper Limit\n";
+    }
+}
+
+
+template <typename test_thres, class Charge_Discharge_SOC_Temp>
+void Early_Warning(test_thres test_treshold, Charge_Discharge_SOC_Temp ClsName) {
+
+    float Warning_tolerance_range = Warning_Tolerance(ClsName);
 
-  if(temp_SOC <= ClsName.Lower_Threshold || temp_SOC >= ClsName.Upper_Threshold) {
+    IsItApproaching_LowerLimit(test_treshold, ClsName, Warning_tolerance_range);
+    IsItApproaching_UpperLimit(test_treshold, ClsName, Warning_tolerance_range);
+
+}
+
+
+template <typename test_thres, class Charge_Discharge_SOC_Temp>
```

Comment:
Good idea to use templates. Templates are an abstraction, so think of a collective name instead of `Charge_Discharge_SOC_Temp`. E.g., `Parameter` or `Telemetry`